### ISSUE TITLE
[Q] How to hide epoll functions which is used in uct/tcp/tcp_ep.c?

### ISSUE THREAD

**Issue by hiroyuki-sato** [openucx/ucx#3652]
## Goal

Now I'm porting OpenUCX to macOS platform.

I would like to use `kqueue` instead of `epoll` on macOS platform.
because it doesn't support that OS.

`epoll` use in `uct/tcp/tcp_ep.c` and `uct/tcp/tcp_iface.c`
This is the issue for discussing maintainer how to hide it

## Background

I introduced new API [`event_set`](https://github.com/openucx/ucx/blob/master/src/ucs/sys/event_set.h) for hiding `epoll` function.
It placed in `ucs/sys/event_set.{h,c}`.(It is **ucs** not in **uct**, by the way)
My first plan was that uses this function instead of `epoll`.

However, `uct_tcp_ep_epoll_ctl` function in `src/uct/tcp/tcp_ep.c` use
`epoll_event.data.ptr` structure. It doesn't support current event_set API.

## Question

* Should I extend `ucs/sys/event_set.{h,c}` for supporting `epoll_event.data.ptr`?
* Or Should I create a new function in `uct`? I'm worried whether It is a good way to use **ucs/event_set** for **uct** layer.
* If the previous answer is to create a new function in `uct` layer, what filename should I use? For example, What do you think to use `src/uct/tcp/tcp_event.c`?

@dmitrygladkov, @yosefe, @shamisp and @brminich  What do you think?

Best regards.

## Reference 

[uct_tcp_ep_epoll_ctl function](https://github.com/openucx/ucx/blob/f96909e0a0e6d7bd007d6d569703e3a8ed12f912/src/uct/tcp/tcp_ep.c#L49-L64)

```C
static void uct_tcp_ep_epoll_ctl(uct_tcp_ep_t *ep, int op)
{
    uct_tcp_iface_t *iface         = ucs_derived_of(ep->super.super.iface,
                                                    uct_tcp_iface_t);
    struct epoll_event epoll_event = {
        .data.ptr                  = ep,
        .events                    = ep->events,
    };
    int ret;


    ret = epoll_ctl(iface->epfd, op, ep->fd, &epoll_event);
    if (ret < 0) {
        ucs_fatal("epoll_ctl(epfd=%d, op=%d, fd=%d) failed: %m",
                  iface->epfd, op, ep->fd);
    }
}
```

epoll structure

```C
 typedef union epoll_data {
     void    *ptr;
     int      fd;
     uint32_t u32;
     uint64_t u64;
 } epoll_data_t;

 struct epoll_event {
     uint32_t     events;    /* Epoll events */
     epoll_data_t data;      /* User data variable */
 };
```

**Comment by dmitrygladkov** [comment]
@hiroyuki-sato Thanks for the information provided in the details!

> * Should I extend `ucs/sys/event_set.{h,c}` for supporting `epoll_event.data.ptr`?

I think this is good to have something like `epoll_event.data.ptr` for UCS event set, because it may be useful for multiple cases (not only for TCP).
So, if we have this in UCS event set API - UCS event set could be used by TCP instead of epoll.

Are you going to implement UCS event set over poll/kqueue for macOS platform? If yes, something like `epoll_event.data.ptr` can be emulated over them

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Thank you for the comment.

I modified event_set implementation to allow user data variables.
`epoll_event.data` is **union** and It doesn't allow set file descriptor and user data at once.
So, I created new data type `ucs_sys_event_set_epoll_data_t` to store it.

https://github.com/hiroyuki-sato/ucx/commit/6a6799ca6873eda220d2c97119404b09aedacc5d

I'll try to replace src/uct/tcp/tcp_ep.c using this function.

P.S.

If this change will merge, the function has two type of user data store. (#3626 and this change)
https://github.com/openucx/ucx/pull/3626/files#diff-7c9812c959527ed11a9d4ea575c0bc51R99
And I'm thinking how to unify it.

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Just sharing information. Maybe you know, UCT require `uct_iface_event_fd_get` implementation and TCP [implementation](https://github.com/openucx/ucx/blob/master/src/uct/tcp/tcp_iface.c#L130-L136) returns fd of epoll.

```C
static ucs_status_t uct_tcp_iface_event_fd_get(uct_iface_h tl_iface, int *fd_p)
{
    uct_tcp_iface_t *iface = ucs_derived_of(tl_iface, uct_tcp_iface_t);


    *fd_p = iface->epfd;
    return UCS_OK;
}
```

I have no idea what descriptor should return in event_set yet.

**Comment by dmitrygladkov** [comment]
@hiroyuki-sato  Hello!

This fd is returned to the user that is used to wait for events using select/poll/epoll/...
So, we need to think about how we should behave in case of we don't have epoll

**Comment by hiroyuki-sato** [comment]
@dmitrygladkov Thank you for the comment!

event_set just wrapping API of kqueue/epoll. It is easy to return file descriptor of epoll/kqueue. 
https://github.com/openucx/ucx/blob/master/src/ucs/sys/event_set.c#L67
But I'm not sure whether it works properly or not **yet**.
What do you think this(just return epoll's file descriptor) implementation is the best way? (At least about epoll)
Should we use other ways? (I'm not sure but it requires returns descriptor like socket/pipe or something?)


This is the event_set implementation using kqueue.
[kqueue(2)](https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2) just return file descriptor. So I expect it works similar epoll_create() function.

https://github.com/hiroyuki-sato/ucx/blob/d0a1d4c37f5a0b702e7aa001716006ecf2b0449b/src/ucs/sys/event_set.c#L86-L101
```C
#ifdef HAVE_SYS_EPOLL_H
    /* Create epoll set the thread will wait on */
    event_set->epfd = epoll_create(1);
    if (event_set->epfd < 0) {
        ucs_error("epoll_create() failed: %m");
        status = UCS_ERR_IO_ERROR;
        goto err_free;
    }
#else
    event_set->epfd = kqueue();
    if ( event_set->epfd == -1 ) {
        ucs_error("kqueue() failed: %m");
        status = UCS_ERR_IO_ERROR;
        goto err_free;
    }
#endif
```

**Comment by shamisp** [comment]
I would return the FD to user as it is. User is responsible to handle this appropriately, based on the underlaying OS. The ```_create``` wrapper looks reasonable.

**Comment by hiroyuki-sato** [comment]
Hello, @shamisp 

Thank you for your advice. I'll implement `_create` wrapper.

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Could you tell me the following questions?
I sometimes confuse the usage of malloc.

Maybe you know, `epoll_ctl` can't use `epoll_event.data.fd` and `epoll_event.data.ptr` at time same time. The reason is that `epoll_event.data` is **union**.

If I understand correctly, It requires to store both(file descriptor and user data) variables into `epoll_event.data.ptr` using a structure for [implementing this idea](https://github.com/openucx/ucx/issues/3652#issuecomment-497835108)

I implemented it the following. I think the comment part require malloc'ed location.
It this correct?

I'm asking this question because current implementation sometime failed. 
And I'm debugging it.

If the answer is correct. It isn't a good place to allocate it.
Because It requires free memory. 
(I'll explain later if the previous question is correct.)

Thank you for your help.

Best regards.

```C
typedef struct ucs_sys_event_set_epoll_data {
    int fd;
    void *udata;
} ucs_sys_event_set_epoll_data_t;

/* snip */

ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
                               ucs_event_set_type_t events, void *udata)
{
    struct epoll_event raw_event;
    int ret;

    /* 
     * Does this variable require to ucs_malloc(), doesn't it?
     * 
     * epoll_event.data is union.
     * So it can't use epoll_event.data.ptr and epoll_event.data.fd at once.
     * That's why I create new data type ucs_sys_event_set_epoll_data_t
     * for storing fd and user data.
     */
    ucs_sys_event_set_epoll_data_t epoll_data; /* Bad code ? */
    // ucs_sys_event_set_epoll_data_t *epoll_data = ucs_malloc..  correct?

    epoll_data.fd = event_fd;
    epoll_data.udata = udata;

    memset(&raw_event, 0, sizeof(raw_event));
    raw_event.events   = ucs_event_set_map_to_raw_events(events);
    raw_event.data.ptr = (void *)&epoll_data;

    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &raw_event);

```

```C
 typedef union epoll_data {
     void        *ptr;
     int          fd;
     uint32_t     u32;
     uint64_t     u64;
 } epoll_data_t;

 struct epoll_event {
     uint32_t     events;      /* Epoll events */
     epoll_data_t data;        /* User data variable */
 };
```

**Comment by dmitrygladkov** [comment]
Hello @hiroyuki-sato 
Thanks for investigating this!

> I [implemented](https://github.com/hiroyuki-sato/ucx/commit/842650c2c48b058f99be0e31bca51097bc1de2b1) it the following. I think the comment part require malloc'ed location.
> It this correct?

Yes, you're right, it requires `malloc()`'ed data and `free` this memory when we read event from the particular fd (`ucs_event_set_wait` - this is the critical path) or in `ucs_event_set_cleanup` if no data were read.

But do you think that we need to store `fd` to the epoll data? Can we avoid it?

I suggest changing `ucs_event_set_handler_t` to accept `epoll_event::data::ptr` instead of `fd`. UCS event_set API user will be responsible to save `fd` inside the user's structure that's passed to `add`/`mod` function. For example, UCT/TCP sets EP to `epoll_event::data::ptr` (TCP EP structure contains `fd` field) and waiting for event usign `epoll_wait` and then just dispatch event usign TCP EP (i.e. just does `recv(ep::fd, ...)` as an example)

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Thank you for your advice. I'll try to implement it.

Does this mean the following? Is this correct?
Maybe, I can imagine it. I'll try it.

Proposal change.

```C
ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
                               ucs_event_set_type_t events, void *callback_data);

typedef void (*ucs_event_set_handler_t)(void *callback_data, int event, void *arg);


ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg)
```

Usage

```C
// event_set user need to create callback data. For example
struct callback_data {
   int fd; 
   void *ptr 
}

// 1. 
// event_set API user register event using this function.
// callback_data require malloce'd memory.
ucs_status_t ucs_event_set_add(event_set, epoll_ctl_fd, UCS_EVENT_SET_EVREAD,
                               (void*)callback_data)

// 1.1
// inside ucs_event_set_add, call epoll_ctl and register callback data.
//
epoll_event.data.ptr = callback_data;
epoll_ctl(epfd, epoll_ctl_fd, &epoll_event)


// 2. 
// event_set user call ucs_event_set_wait function with callback function.
ucs_event_set_wait(event_set, callback_function, args)

// 2.1
// inside ucs_event_set_wait
// call epoll_wait function. 
// callback data is stored inside events[i].data.ptr
epoll_wait(epfd, &events, MAX_EVENTS, TIMEOUT)

// 2.2
// Call callback function with callback_data
callback_data = events.data.ptr.
callback_function(callback_data, event_type, args);

// 3 
// inside callback function
callback_function(void *callback_data, int event, void *args){
    read(callback_data->fd,...);
    do_somthing_function(callback_data->ptr,..)
    // ....
}
```

**Comment by dmitrygladkov** [comment]
> Does this mean the following? Is this correct?
> Maybe, I can imagine it. I'll try it.
> 
> Proposal change.

Hello @hiroyuki-sato 

Yes, that's what I meant
Thank you!

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 
Thank you for your help! I'll try it. 
Please wait for update.

**Comment by dmitrygladkov** [comment]
> Thank you for your help! I'll try it.
> Please wait for update.

@hiroyuki-sato Thank you for great efforts!

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

I modified event_set API for using callback data.
https://github.com/hiroyuki-sato/ucx/commit/6ab14f091e250a30cec436657d86ef8b53d58bfe

Next, I'll modify ucs/async/thread.c and uct/tcp using this API.
Modification ucs/async/thread.c will based on [this change](https://github.com/openucx/ucx/pull/3610/files).
I'm thinking about how to free malloc'ed memory.

Stay tuned for an update.

Best regards.

**Comment by dmitrygladkov** [comment]
Hello @hiroyuki-sato 

> I modified event_set API for using callback data.
> [hiroyuki-sato@6ab14f0](https://github.com/hiroyuki-sato/ucx/commit/6ab14f091e250a30cec436657d86ef8b53d58bfe)

I went through your changes, they LGTM.

> I'm thinking about how to free malloc'ed memory.

Where free `malloc()'ed memory is required?

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Thank you for the comment.

I don't completely understand ucs/async/thread.c yet.
Maybe the answer is simple. 

The following codes copied from this PR.
https://github.com/openucx/ucx/pull/3610/files

This callback function changed `int fd` to `void *callback_data`.
So We need to get `fd` using an alternative way.

https://github.com/openucx/ucx/blob/8fcadeda04fb5e70f481e48854e4cfb67153faa5/src/ucs/async/thread.c#L65-L82

This `fd` was set in the following parts using `ucs_event_set_add` function.

https://github.com/openucx/ucx/blob/8fcadeda04fb5e70f481e48854e4cfb67153faa5/src/ucs/async/thread.c#L284-L300


`ucs_event_set_add` added `void *callback_data` like

```
ucs_event_set_add(thread->event_set, event_fd, UCS_EVENT_SET_EVREAD, callback_data)
```

I think that I need to store `fd` inside `callback_data`.  How to it?

**Comment by dmitrygladkov** [comment]
Hello @hiroyuki-sato 

I've got the issue nwo

> I think that I need to store `fd` inside `callback_data`. How to it?

as `callback_data` is 8-byte storage, we can set `fd` w/o allocating additional memory - just pack the `fd` in the following way:
```
callback_data = (uintptr_t)fd;
```
and the unpack it:
```
fd = (int)(uintptr_t)callback_data;
```

Hope it should solve the issue.

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Thank you for your great advice!.

I modified `ucs/async/thread.c` using your proposal.
https://github.com/hiroyuki-sato/ucx/commit/bbfa9bc091d48f4cede671d8d8cf5f297a92d363

I needed `(void *)(uintptr_t)fd` syntax for avoiding compiler error.

Next, I'll change `uct/tcp`.

**Comment by dmitrygladkov** [comment]
Hello @hiroyuki-sato 

> I needed `(void *)(uintptr_t)fd` syntax for avoiding compiler error.

Yes, right, this is needed

Are you going to create a separate PR with the `uct/tcp` changes or combine it with the `ucs/async` changes?

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Thank you for helping me every time!

Yes. I'll try to change uct/tcp on another branch. If that works well, I'll create three PRs.

* extend event_set
* use event_set in `ucs/async/thread.c` (Force push #3610)
* use event_set in `uct/tcp`

Best regards.

**Comment by dmitrygladkov** [comment]
Hello @hiroyuki-sato 

> Yes. I'll try to change uct/tcp on another branch. If that works well, I'll create three PRs.

Thank you! This sounds like a great plan to me

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

I modified uct/tcp for using event_set API.
https://github.com/hiroyuki-sato/ucx/commit/2b3975af9c6769e501438b3138d6134ef5fc160e

But It's still buggy. 

I'm getting the following error.

```
[xxxx:45599:0:45599]      tcp_ep.c:327  Fatal: ucs_event_set_del(fd=10) failed: No such file or directory
[xxxx:45599:0:45599] Process frozen, press Enter to attach a debugger...
```


Anyway, I'm thinking about the following. Could you tell me your thought?
`uct_tcp_iface_progress` function is a difficult part using event_set API.

Best regards.

## Issues

* How to count `count` and `nevents` in handler
* Does ucs_event_set_wait return UCS_ERR_BUSY(or something) if `errno == EINTR?`
* Does `UCT_TCP_MAX_EVENTS` is important?
`
### How to count `count` and `nevents` in handler

What do you think about this implementation? I used a callback trick again.

https://github.com/hiroyuki-sato/ucx/blob/2b3975af9c6769e501438b3138d6134ef5fc160e/src/uct/tcp/tcp_iface.c#L170-L172

```C
status = ucs_event_set_wait(iface->event_set, 0,
                            uct_tcp_iface_handle_events, 
                            cb_arg);
```

### Does ucs_event_set_wait should return UCS_ERR_BUSY(or something) if `errno == EINTR?`

Current(epoll) implementation detect `EINTR`. 

https://github.com/hiroyuki-sato/ucx/blob/2b3975af9c6769e501438b3138d6134ef5fc160e/src/uct/tcp/tcp_iface.c#L178-L184

```
if (ucs_unlikely((nevents < 0))) {
    if (errno == EINTR) {
        /* force a new loop iteration */
        nevents = max_events;
        continue;
    }
```

But current ucs_event_set_wait doesn't care `EINTR`.

https://github.com/hiroyuki-sato/ucx/blob/2b3975af9c6769e501438b3138d6134ef5fc160e/src/ucs/sys/event_set.c#L149-L162

```C
    nready = epoll_wait(event_set->epfd, ep_events, UCS_EVENT_EPOLL_MAX_EVENTS,
                        timeout_ms);
    if ((nready < 0) && (errno != EINTR)) {
        ucs_error("epoll_wait() failed: %m");
        return UCS_ERR_IO_ERROR;
    }
    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                   event_set->epfd, timeout_ms, nready);


    for (i=0; i < nready; i++) {
        int events = ucs_event_set_map_to_events(ep_events[i].events);
        event_set_handler(ep_events[i].data.ptr, events, arg);
    }
    return UCS_OK;
```


Maybe, EINTR can detect the following way

```
 if ( status == UCS_OK && nevents == 0 )
```

### Does `UCT_TCP_MAX_EVENTS` is important?

event_set always wait UCS_EVENT_EPOLL_MAX_EVENTS times.

Current implementation is changing how many events wait in epoll_wait 


What do you think to define `UCT_TCP_MAX_EVENTS` like the following?

```C
/* Currently both value are 16 */
#define UCT_TCP_MAX_EVENTS UCS_EVENT_EPOLL_MAX_EVENTS
```

Current event_set API has `UCS_EVENT_EPOLL_MAX_EVENTS`

https://github.com/hiroyuki-sato/ucx/blob/2b3975af9c6769e501438b3138d6134ef5fc160e/src/uct/tcp/tcp_iface.c#L178-L179

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

I updated codes for fixing errors.
Current implementation passed tests.
(It may need to improve `uct_tcp_iface_progress` function)
https://github.com/hiroyuki-sato/ucx/commit/ef945bc1f54d86bf2363f6aa7b94b7390ac65dd0

If uct/tcp require custom epoll_wait size(`iface->config.max_poll`),
What do you think extend event_set API like `ucs_event_set_create(.., max_poll_size)`?

Best regards

**Comment by dmitrygladkov** [comment]
Hello @hiroyuki-sato 

> I modified uct/tcp for using event_set API.
> hiroyuki-sato@2b3975a
> But It's still buggy.
> I'm getting the following error.

Gave comments on the code that should fix the bug

> ### How to count `count` and `nevents` in handler
> What do you think about this implementation? I used a callback trick again.

Yes, this is a good way to do this

> Does ucs_event_set_wait should return UCS_ERR_BUSY(or something) if `errno == EINTR?`


> Maybe, EINTR can detect the following way
> 
> ```
>  if ( status == UCS_OK && nevents == 0 )
> ```

we should continue sleeping(waiting) for events in epoll w/o involving UCS event_set API user into this process.
```
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg)
{
    int nready;
    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
    int i;

    if (event_set_handler == NULL) {
        return UCS_ERR_INVALID_PARAM;
    }

    for (;;) {
        nready = epoll_wait(event_set->epfd, ep_events,
                            UCS_EVENT_EPOLL_MAX_EVENTS,
                            timeout_ms);
        if (nready < 0) {
            if (errno == EINTR) {
                /* here we should take care about timeout_ms (i.e. decrease it),
                 * but it is ok to not od this */
                continue;
            }
            ucs_error("epoll_wait() failed: %m");
            return UCS_ERR_IO_ERROR;
        }
    }
    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                   event_set->epfd, timeout_ms, nready);

    for (i=0; i < nready; i++) {
        int events = ucs_event_set_map_to_events(ep_events[i].events);
        event_set_handler(ep_events[i].data.ptr, events, arg);
    }
    return UCS_OK;
}
```


> Does `UCT_TCP_MAX_EVENTS` is important?

it is used to declare an array of events on the stack, not it's not important
I think the more important to keep current TCP progress engine as is, i.e. handling not more than the requested number of events.
i.e. `ucs_event_set_wait` should be updated in the following way to accept a new parameter `max_events` that will be used under the hood by `ucs_event_set_wait` to handle events <= `max_events`:
```
unsigned ucs_event_set_wait(ucs_sys_event_set_t *event_set, int max_events,
                            int timeout_ms,
                            ucs_event_set_handler_t event_set_handler,
                            void *arg)
{
    int read_events = 0;
    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
    int i, max_wait_events, nready;

    ucs_assert(event_set_handler != NULL);

    do {
        max_wait_events = ucs_min(max_events - read_events,
                                  UCS_EVENT_EPOLL_MAX_EVENTS);
        nready = epoll_wait(event_set->epfd, ep_events,
                            max_wait_events, timeout_ms);
        // handle errors here as discussed above...
        for (i = 0; i < nready; i++) {
            int events = ucs_event_set_map_to_events(ep_events[i].events);
            event_set_handler(ep_events[i].data.ptr, events, arg);
        }

        read_events += nready;
    } while ((read_events < iface->config.max_poll) && (nready == max_wait_events));

    return read_events;
}
```
btw, change to return the number of events read as a user can use it and status is always UCS_OK in the current implementation

> What do you think to define `UCT_TCP_MAX_EVENTS` like the following?

I think there is no need for this constant anymore


Thank you!

**Comment by dmitrygladkov** [comment]
> If uct/tcp require custom epoll_wait size(`iface->config.max_poll`),
> What do you think extend event_set API like `ucs_event_set_create(.., max_poll_size)`?

I think we can go with the idea that I recommend you in the comment above - it looks more flexible, because a user may decide to pass a different number of polled events depending on different situations

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov 

Thank you for your great help!
I almost modified codes. 
I think It's time to create new PR. Let's discuss in a new PR. 
I created #3682 for summaries those works.

**Comment by dmitrygladkov** [comment]
> Hello, @dmitrygladkov
> 
> Thank you for your great help!
> I almost modified codes.
> I think It's time to create new PR. Let's discuss in a new PR.
> I created #3682 for summaries those works.

Hello @hiroyuki-sato 

Thank you for working on this!!!
